File: algorithm/34.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> searchRange(vector<int>& nums, int target);

private:
    int searchLeft(vector<int>& nums, int target);
    int searchRight(vector<int>& nums, int target);
};

vector<int> Solution::searchRange(vector<int>& nums, int target)
{
    if(nums.empty()) return {-1, -1};

    int n = nums.size();
    vector<int> res = {-1,-1};

    res[0] = searchLeft(nums, target);
    res[1] = searchRight(nums, target);
    return res;
}

int Solution::searchLeft(vector<int>& nums, int target) //左边界是找第一个等于target的元素
{
    int n = nums.size();
    int low=0, high=n-1;
    while (low<=high)
    {
        int mid = low + ((high-low)>>1);
        if (nums[mid]>=target)
        {
            if(nums[mid]==target && (mid==0 || nums[mid-1]<target)) return mid;
            else high = mid-1;
        }
        else low = mid+1;
    }
    return -1;
}

int Solution::searchRight(vector<int>& nums, int target) //右边界找最后一个等于target的元素
{
    int n = nums.size();
    int low=0, high=n-1;
    while (low<=high)
    {
        int mid = low + ((high-low)>>1);
        if(nums[mid]<=target)
        {
            if(nums[mid]==target && (mid==n-1 || nums[mid+1]>target)) return mid;
            else low = mid+1;
        }
        else high = mid-1;
    }
    return -1;
}

int main()
{
    vector<int> nums = {5,7,7,8,8,10};
    int target = 6;
    Solution obj;
    vector<int> result = obj.searchRange(nums, target);
    return 0;
}

File: algorithm/go/firstround/34.go
package firstround

func searchRange(nums []int, target int) []int {
	left := findLeftBound(nums, target)
	right := findRightBound(nums, target)
	return []int{left, right}
}

// 寻找左侧边界
func findLeftBound(nums []int, target int) int {
	left, right := 0, len(nums)-1

	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] == target {
			right = mid-1
		} else if nums[mid] > target {
			right = mid-1
		} else if nums[mid] < target {
			left = mid+1
		}
	}

	// 判断left是否越界或者nums[left]是否等于target
	if left < 0 || left >= len(nums) || nums[left] != target {
		return -1
	}
	return left
}

// 寻找右侧边界
func findRightBound(nums []int, target int) int {
	left, right := 0, len(nums)-1
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] == target {
			left = mid+1
		} else if nums[mid] < target {
			left = mid+1
		} else if nums[mid] > target {
			right = mid-1
		}
	}

	// 判断right是否越界或者nums[right]是否等于target
	if right < 0 || right >= len(nums) || nums[right] != target {
		return -1
	}
	return right
}
